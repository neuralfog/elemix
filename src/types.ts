import { type HtmlTemplate, html as rhtml } from '@neuralfog/elemix-renderer';

export type Template = HtmlTemplate;
export const html = rhtml;

export const RenderTrigger = {
    PROPS: 'PROPS',
    SIGNAL: 'SIGNAL',
    LOCAL_STATE: 'LOCAL_STATE',
    ON_MOUNT: 'ON_MOUNT',
} as const;

export type RenderTriggerType =
    | (typeof RenderTrigger)[keyof typeof RenderTrigger]
    | string;
