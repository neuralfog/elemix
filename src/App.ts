type AppConfig = {
    cssReset?: string;
    entryPoint?: () => Promise<any>;
};

// biome-ignore lint:
export class App {
    static config: AppConfig = {};

    public static mergeConfigs(config: AppConfig): void {
        App.config = {
            ...App.config,
            ...config,
        };
    }
}

export const initApp = (config: AppConfig): void => {
    App.mergeConfigs(config);
    void App.config.entryPoint?.();
};
