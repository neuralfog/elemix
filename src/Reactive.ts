import type { Component } from './component/Component';
import type { RenderTriggerType } from './types';

export class Reactive<State> {
    public subscribers = new Set<Component>();
    private proxy: State;
    private proxySet = new WeakSet<any>();

    public get value(): State {
        return this.proxy;
    }

    constructor(
        state: any,
        private renderTrigger?: RenderTriggerType,
    ) {
        this.proxy = this.create(state);
    }

    public subscribe(instance: Component): Reactive<State> {
        if (!this.subscribers.has(instance)) this.subscribers.add(instance);
        return this;
    }

    public unsubscribe(instance: Component): Reactive<State> {
        this.subscribers.delete(instance);
        return this;
    }

    private create(state: any): State {
        const reactive = this;

        return new Proxy(state, {
            get(target, name): unknown {
                const prop = target[name];

                unsupportedCollectionError(prop);

                if (
                    typeof prop === 'object' &&
                    prop !== null &&
                    !reactive.proxySet.has(prop)
                ) {
                    const propRef = new Proxy(prop, this);
                    reactive.proxySet.add(propRef);
                    return propRef;
                }

                return prop;
            },
            set(target, name, value): boolean {
                target[name] = value;

                unsupportedCollectionError(target);

                if (Array.isArray(target) && name === 'length') {
                    reactive.notify();
                    return true;
                }

                reactive.notify();
                return true;
            },
        }) as State;
    }

    private notify(): void {
        for (const subscriber of this.subscribers) {
            subscriber.render(this.renderTrigger);
        }
    }
}

export const UNSUPPORTED_COLLECTION_ERROR_MESSAGE =
    'Reactive state does not support collections: Map, WeakMap, Set, WeakSet';

const unsupportedCollectionError = (prop: any): any => {
    if (
        prop instanceof Map ||
        prop instanceof WeakMap ||
        prop instanceof Set ||
        prop instanceof WeakSet
    ) {
        throw new Error(UNSUPPORTED_COLLECTION_ERROR_MESSAGE);
    }
};
