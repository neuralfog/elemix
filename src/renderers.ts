import type { Renderer } from './component/Renderer';

export const activeRenderers = new Set<Renderer>();
