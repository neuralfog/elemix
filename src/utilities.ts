export const ref = <Value>(
    value: Value | undefined = undefined,
): { value?: Value } => {
    return {
        value,
    };
};
