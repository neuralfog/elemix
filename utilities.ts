export { mergeClasses } from '@neuralfog/elemix-renderer/utils';
export * from './src/utilities';
