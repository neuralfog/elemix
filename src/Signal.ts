import { Reactive } from './Reactive';
import { RenderTrigger } from './types';

export const signal = <State>(
    state: State,
    renderTrigger?: string,
): Reactive<State> =>
    new Reactive(state, renderTrigger || RenderTrigger.SIGNAL);
