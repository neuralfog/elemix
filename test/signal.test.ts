import { expect, test, describe, beforeEach, vi } from 'vitest';
import { HTML } from '@neuralfog/elemix-testing/snapshots';
import { html } from '@neuralfog/elemix-renderer';
import { Present } from '@neuralfog/elemix-testing';
import type { MainApp } from './fixtures/MainApp';
import { restoreSignal, store } from './fixtures/signal';
import { RenderTrigger } from '../src/types';
import type { TestComp } from './fixtures/TestComp';
import type { ListApp } from './fixtures/ListApp';
import { render } from '../utilities';

import './fixtures/ListApp';
import './fixtures/MainApp';

describe('Signal', () => {
    beforeEach(() => {
        document.body.innerHTML = '';
        restoreSignal();
    });

    test('Initial State', async () => {
        const presenter = new Present().screen(html`<main-app />`);

        await render();

        expect(HTML(presenter.root<MainApp>())).toMatchSnapshot();
    });

    test('Initial State - Single Component', async () => {
        const presenter = new Present().screen(html`<list-app />`);

        await render();

        expect(HTML(presenter.root<ListApp>())).toMatchSnapshot();
    });

    test('Initial State - Multiple Component', async () => {
        const presenter = new Present().screen(html`<list-app />`);
        const listApp = presenter.root<ListApp>();
        listApp.state.count = 5;

        await render();

        expect(HTML(presenter.root<ListApp>())).toMatchSnapshot();
    });

    test('Modify Signal', async () => {
        const presenter = new Present().screen(html`<main-app />`);

        await render();

        const testComp = presenter.getByTag<TestComp>('test-comp');
        const onRender = vi.spyOn(testComp, 'onRender');

        expect(store.value.value).toBe('Initial Signal Value');
        store.value.value = 'New Value';

        await render();

        expect(HTML(presenter.root<MainApp>())).toMatchSnapshot();
        expect(onRender).toHaveBeenCalledOnce();
        expect(onRender).toBeCalledWith([RenderTrigger.SIGNAL]);
    });
});

describe('Signal Cleanup', () => {
    beforeEach(() => {
        document.body.innerHTML = '';
        restoreSignal();
    });

    test('Single Component', async () => {
        const presenter = new Present().screen(html`<list-app />`);

        await render();

        expect(store.subscribers.size).toBe(1);

        const listApp = presenter.root<ListApp>();
        listApp.state.count = 0;

        await render();

        expect(store.subscribers.size).toBe(0);
    });

    test('Multiple Component', async () => {
        const presenter = new Present().screen(html`<list-app />`);
        const listApp = presenter.root<ListApp>();
        listApp.state.count = 5;

        await render();

        expect(store.subscribers.size).toBe(5);

        listApp.state.count = 0;

        await render();

        expect(store.subscribers.size).toBe(0);
    });
});
