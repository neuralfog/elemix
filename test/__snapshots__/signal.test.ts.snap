// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Signal > Initial State - Multiple Component 1`] = `
"
<list-app><!-- #shadowroot -->
  <h1>ListApp</h1>
  <ul>
    <li>
      <test-comp><!-- #shadowroot -->
        <div>
          <h1>TestComp</h1>
          <p id="test-comp-state">Initial State Value</p>
          <p id="test-comp-state-2">0</p>
          <p id="test-comp-signal">Initial Signal Value</p>
          <p id="test-comp-props">-</p>
        </div>
      </test-comp>
    </li>
    <li>
      <test-comp><!-- #shadowroot -->
        <div>
          <h1>TestComp</h1>
          <p id="test-comp-state">Initial State Value</p>
          <p id="test-comp-state-2">0</p>
          <p id="test-comp-signal">Initial Signal Value</p>
          <p id="test-comp-props">-</p>
        </div>
      </test-comp>
    </li>
    <li>
      <test-comp><!-- #shadowroot -->
        <div>
          <h1>TestComp</h1>
          <p id="test-comp-state">Initial State Value</p>
          <p id="test-comp-state-2">0</p>
          <p id="test-comp-signal">Initial Signal Value</p>
          <p id="test-comp-props">-</p>
        </div>
      </test-comp>
    </li>
    <li>
      <test-comp><!-- #shadowroot -->
        <div>
          <h1>TestComp</h1>
          <p id="test-comp-state">Initial State Value</p>
          <p id="test-comp-state-2">0</p>
          <p id="test-comp-signal">Initial Signal Value</p>
          <p id="test-comp-props">-</p>
        </div>
      </test-comp>
    </li>
    <li>
      <test-comp><!-- #shadowroot -->
        <div>
          <h1>TestComp</h1>
          <p id="test-comp-state">Initial State Value</p>
          <p id="test-comp-state-2">0</p>
          <p id="test-comp-signal">Initial Signal Value</p>
          <p id="test-comp-props">-</p>
        </div>
      </test-comp>
    </li>
  </ul>
</list-app>
"
`;

exports[`Signal > Initial State - Single Component 1`] = `
"
<list-app><!-- #shadowroot -->
  <h1>ListApp</h1>
  <ul>
    <li>
      <test-comp><!-- #shadowroot -->
        <div>
          <h1>TestComp</h1>
          <p id="test-comp-state">Initial State Value</p>
          <p id="test-comp-state-2">0</p>
          <p id="test-comp-signal">Initial Signal Value</p>
          <p id="test-comp-props">-</p>
        </div>
      </test-comp>
    </li>
  </ul>
</list-app>
"
`;

exports[`Signal > Initial State 1`] = `
"
<main-app><!-- #shadowroot -->
  <h1>MainApp</h1>
  <p id="main-app-state">Initial State Value</p>
  <test-comp><!-- #shadowroot -->
    <div>
      <h1>TestComp</h1>
      <p id="test-comp-state">Initial State Value</p>
      <p id="test-comp-state-2">0</p>
      <p id="test-comp-signal">Initial Signal Value</p>
      <p id="test-comp-props">Initial Color - 0</p>
    </div>
  </test-comp>
</main-app>
"
`;

exports[`Signal > Modify Signal 1`] = `
"
<main-app><!-- #shadowroot -->
  <h1>MainApp</h1>
  <p id="main-app-state">Initial State Value</p>
  <test-comp><!-- #shadowroot -->
    <div>
      <h1>TestComp</h1>
      <p id="test-comp-state">Initial State Value</p>
      <p id="test-comp-state-2">0</p>
      <p id="test-comp-signal">New Value</p>
      <p id="test-comp-props">Initial Color - 0</p>
    </div>
  </test-comp>
</main-app>
"
`;
